File: 프로그래머스/3/60062. 외벽 점검/외벽 점검.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(int n, vector<int> weak, vector<int> dist) {
    int answer = 1e9;
    
    // weak 길이 확장
    int weak_len = weak.size();
    for(int i = 0 ; i < weak_len ; i++)
        weak.push_back(weak[i] + n);
    
    // weak 첫번째부터 순회
    for(int i = 0 ; i < weak_len ; i++) {
        
        // dist 순열
        do{
            int cnt = 1;
            int limit_dist = weak[i] + dist[cnt - 1];
            
            for(int j = i ; j < i + weak_len ; j++) {
                // 새로운 사람 투입
                if(limit_dist < weak[j]){
                    cnt++;
                    if(cnt > dist.size())
                        break;
                    else
                        limit_dist = weak[j] + dist[cnt - 1];
                }
                
            }
            
            // 사람 체크
            if(cnt <= dist.size())
                answer = min(answer, cnt);
            
        }while(next_permutation(dist.begin(), dist.end()));
        
    }
    
    
    if(answer == 1e9)
        answer = -1;
    
    return answer;
}

File: 프로그래머스/3/60062. 외벽 점검/외벽 점검.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int n, int[] weak, int[] dist) {
        
        // 원형 이어 붙이기
        List<Integer> newWeak = Arrays.stream(weak).boxed().map(i -> i + n)
            .collect(Collectors.toList());
        
        List<Integer> weakList = Arrays.stream(weak).boxed().collect(Collectors.toList());
        weakList.addAll(newWeak);
        
        // dist 순열
        List<int[]> distPermutation = new ArrayList<>();
        getPermutation(distPermutation, dist, new boolean[dist.length], new ArrayDeque<>());
        
        int ret = Integer.MAX_VALUE;
        
        for(int i = 0 ; i < weak.length ; i++){
            int start = weakList.get(i);
            
            // 시작점부터 순열
            for(int[] per : distPermutation){
                int idx = 0;
                int cnt = 0;
                int need = start;
                // 점검
                for(int j = i ; j < i + weak.length ; j++){
                    
                    // 점검 완료된 것들
                    if(need > weakList.get(j)){
                        cnt++;
                        continue;
                    }
                    
                    // 친구 투입
                    if(idx < per.length){
                        need = weakList.get(j) + per[idx] + 1;
                        idx++;
                        cnt++;
                    }
                }
                
                if(cnt == weak.length && idx <= per.length){
                    ret = Math.min(ret, idx);
                }
            }
        }
        
        return ret == Integer.MAX_VALUE ? -1 : ret;
    }
    
    void getPermutation(List<int[]> permutation, int[] dist, boolean[] visited, Deque<Integer> res){
        if(dist.length == res.size()){
            permutation.add(res.stream().mapToInt(Integer::valueOf).toArray());
            return;
        }
        
        for(int i = 0 ; i < dist.length ; i++){
            if(!visited[i]){
                visited[i] = true;
                res.offerLast(dist[i]);
                getPermutation(permutation, dist, visited, res);
                res.pollLast();
                visited[i] = false;
            }
        }
    }
}